pandas/_libs/algos.pyx

```diff
@@ -1231,15 +1237,31 @@ def diff_2d(
                         start, stop = 0, sx + periods
                     for j in range(sy):
                         for i in range(start, stop):
-                            out[i, j] = arr[i, j] - arr[i - periods, j]
+                            left = arr[i, j]
+                            right = arr[i - periods, j]
+                            if out_t is int64_t and datetimelike:
+                                if left == NPY_NAT or right == NPY_NAT:
+                                    out[i, j] = NPY_NAT
+                                else:
+                                    out[i, j] = left - right
+                            else:
+                                out[i, j] = left - right
                 else:
                     if periods >= 0:
                         start, stop = periods, sy
                     else:
                         start, stop = 0, sy + periods
                     for j in range(start, stop):
                         for i in range(sx):
-                            out[i, j] = arr[i, j] - arr[i, j - periods]
+                            left = arr[i, j]
```

Comment:
rather than repeating can you make an in line function?

Reply:
i wasnt able to, no.  the trouble is we dont have a good way to declare the return type